### PR TITLE
Add parentheses when converting Substring to Range

### LINKED ISSUE #36997: IDE0057 / "Use range operator": incorrect code change (missing parentheses)
Actual behaviour:
In code like the following:
```
string version = ...;
version = version. Substring(1, (version. IndexOf("-") == -1 ? version. Length : version. IndexOf("-")) - 1)
```
Visual Studio 2019 suggests to &quot;Use range operator&quot;, &quot;IDE0057 Substring can be simplified&quot;, and when invoked, changes the second line to:

```
version = version[1..version. IndexOf("-") == -1 ? version. Length : version. IndexOf("-")];
```
This results in compiler error CS0019: &quot;Operator '==' cannot be applied to operands of type 'Range' and 'int'&quot;, which is correct.

Expected behaviour:
The range operator ``.. `` apparently has higher precedence than the ternary conditional operator ``?:``, so parentheses should have been inserted as appropriate (or rather, in my example, been kept as they were); in this case, the second line might have instead been converted to:
```
version = version[1..( version. IndexOf("-") == -1 ? version. Length : version. IndexOf("-"))];
```

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/633621/ide0057-use-range-operator-incorrect-code-change-m.html
VSTS ticketId: 942430_
_These are the original issue comments:_

Visual Studio Feedback System on 7/5/2019, 00:57 AM (39 min ago): <markdown><p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>
</markdown>
_These are the original issue solutions:_
(no solutions)

### PR DESCRIPTION
**Pull request by meziantou**
resolve #36997

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 1]
```diff
@@ -272,6 +272,14 @@ public static bool CanRemoveParentheses(this ParenthesizedExpressionSyntax node,
                 return true;
             }
 
+            // Range expression
+            // x[(y)..] => x[y..]
+            if (node.Parent is RangeExpressionSyntax)
+            {
+                if (expression.IsKind(SyntaxKind.ElementAccessExpression))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs:279]: this doesn't seem right.  i think the appropriate question would be is the expr in teh parens higher precedence than the rangeepression?  if so, parens can be removed.  if not, they have to say.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs [hunk 2]
```diff
@@ -272,6 +272,14 @@ public static bool CanRemoveParentheses(this ParenthesizedExpressionSyntax node,
...
+                    return true;
+            }
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs:281]: ❔ Why is this getting added? It seems like the call to `RemovalChangesAssociation` would already be responsible for handling this case.

Comment (meziantou) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs:281]: I was expecting a comment on this one :)

The current code doesn't remove the parentheses in the following case `s[(a[0])..]`. `RemovalChangesAssociation` is conservative as `RangeExpressionSyntax` precedence is `None`. I think I need to add `OperatorPrecedence.Range` between `Multiplicative` and `Unary` ([doc](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/#operator-precedence)) to fix the issue.

In `GetOperatorPrecedence` there are comments that seems to come from the C# spec (`// From C# spec, 7.3.1:`). Where is the URL of this spec so I can add a similar comment for range?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs:281]: >  I think I need to add OperatorPrecedence.Range between Multiplicative and Unary (doc) to fix the issue.

Yes, that would make sense to me.  ranges are here: https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/ranges.md

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ExpressionSyntaxExtensions.cs [hunk 3]
```diff
@@ -700,6 +700,9 @@ public static OperatorPrecedence GetOperatorPrecedence(this ExpressionSyntax exp
 
                     return OperatorPrecedence.Unary;
 
+                case SyntaxKind.RangeExpression:
+                    return OperatorPrecedence.Range;
```
Comment (meziantou) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ExpressionSyntaxExtensions.cs:704]: Do I need to add a comment such as

````
                    // From C# spec, https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/ranges.md#systemrange
                    // Range: ..
````

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ExpressionSyntaxExtensions.cs:704]: that would be nice.  but not necessary.

**Pull request merged**